File: server/util/hisutil/crypt/Crypto.cpp
#include "stdafx.h"
#include "Crypto.h"

namespace HisUtil {

	const UINT32 C1 = 42875;
	const UINT32 C2 = 26119;
	const UINT32 KEY = 71957;

	BOOL CCrypto::Encrypt(unsigned char *source, UINT32 length, unsigned char *destination)
	{
		UINT32 Key = KEY;

		if (!source || !destination || length <= 0)
			return FALSE;

		for (UINT32 i = 0; i < length; ++i)
		{
			destination[i] = source[i] ^ (Key >> 8);
			
			//Encrypt된 값으로 생성되는 Key는
			//동일한 순서를 갖는다.
			Key = (destination[i] + Key) * C1 + C2;
			//printf("%d ", Key);
		}
		//printf("\n");

		return TRUE;
	}

	BOOL CCrypto::Decrypt(unsigned char *source, UINT32 length, unsigned char *destination)
	{
		unsigned char PreviousBlock = 0;
		UINT32 Key = KEY;

		if (!source || !destination || length <= 0)
			return FALSE;

		for (UINT32 i = 0; i < length; ++i)
		{
			PreviousBlock = source[i];
			destination[i] = source[i] ^ (Key >> 8);
			
			//Encrypt된 값으로 생성되는 Key는
			//동일한 순서를 갖는다.
			Key = (PreviousBlock + Key) * C1 + C2;
			//printf("%d ", Key);
		}
		//printf("\n");

		return TRUE;
	}

};//..namespace HisUtil {

File: FreeNet_v0.1/FreeNet/Crypto.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace FreeNet
{
public class Crypto
{
    static UInt32 C2 = 26119;
    static UInt32 KEY = 71957;
    static UInt32 C1 = 42875;

    static public bool Encrypt(byte[] source, UInt32 sourceIndex, ref byte[] destination, UInt32 destinationIndex, UInt32 length)
    {
        UInt32 key = KEY;

        if (length <= 0) return false;

        for (UInt32 i = 0; i < length; ++i)
        {
            destination[destinationIndex + i] = (byte)(source[sourceIndex + i] ^ (key >> 8));
            key = (destination[destinationIndex + i] + key) * C1 + C2;
        }
        return true;
    }
    static public bool Decrypt(byte[] source, UInt32 sourceIndex, ref byte[] destination, UInt32 destinationIndex, UInt32 length)
    {
        byte PreviousBlock = 0;
        UInt32 key = KEY;

        if (length <= 0) return false;

        for (UInt32 i = 0; i < length; ++i)
        {
            PreviousBlock = source[sourceIndex + i];
            destination[destinationIndex + i] = (byte)(source[sourceIndex + i] ^ (key >> 8));

            key = (PreviousBlock + key) * C1 + C2;
        }
        return true;
    }
}
}
